src/category_theory/sites/subsheaf.lean

```diff
@@ -0,0 +1,236 @@
+/-
+Copyright (c) 2022 Andrew Yang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Andrew Yang
+-/
+import category_theory.elementwise
+import category_theory.sites.compatible_sheafification
+import category_theory.limits.constructions.epi_mono
+
+/-!
+
+# Subsheaf of types
+
+We define the sub(pre)sheaf of a type valued presheaf.
+
+## Main results
+
+- `category_theory.grothendieck_topology.subpresheaf` :
+  A subpresheaf of a presheaf of types.
+- `category_theory.grothendieck_topology.subpresheaf.sheafify` :
+  The sheafification of a subpresheaf as a subpresheaf. Note that this is a sheaf only when the
+  whole sheaf is.
+- `category_theory.grothendieck_topology.subpresheaf.sheafify_is_sheaf` :
+  The sheafification is a sheaf
+- `category_theory.grothendieck_topology.subpresheaf.sheafify_lift` :
+  The descent of a map into a sheaf to the sheafification.
+-/
+
+universes w v u
+
+open opposite category_theory
+
+namespace category_theory.grothendieck_topology
+
+variables {C : Type u} [category.{v} C] (J : grothendieck_topology C)
+
+/-- The subpresheaf of a presheaf. -/
```

Comment:
To me "The subpresheaf" suggests that every presheaf has a unique subpresheaf. Maybe rephrase as
```suggestion
/-- A subpresheaf of a presheaf consists of a subset of `F.obj U` for every `U`,
compatible with the restriction maps `F.map i`. -/
```